pkg/sources/github/github.go

```diff
@@ -824,53 +825,70 @@ func (s *Source) scan(ctx context.Context, installationClient *github.Client, ch
 	return nil
 }
 
+var (
+	rateLimitMu         sync.RWMutex
+	rateLimitResumeTime time.Time
+)
+
 // handleRateLimit returns true if a rate limit was handled
-// Unauthenticated access to most github endpoints has a rate limit of 60 requests per hour.
-// This will likely only be exhausted if many users/orgs are scanned without auth
+//
+// Unauthenticated users have a rate limit of 60 requests per hour.
+// Authenticated users have a rate limit of 5,000 requests per hour,
+// however, certain actions are subject to a stricter "secondary" limit.
+// https://docs.github.com/en/rest/overview/rate-limits-for-the-rest-api
 func (s *Source) handleRateLimit(errIn error, res *github.Response) bool {
-	var (
-		knownWait  = true
-		remaining  = 0
-		retryAfter time.Duration
-	)
-
-	// GitHub has both primary (RateLimit) and secondary (AbuseRateLimit) errors.
-	var rateLimit *github.RateLimitError
-	var abuseLimit *github.AbuseRateLimitError
-	if errors.As(errIn, &rateLimit) {
-		// Do nothing
-	} else if errors.As(errIn, &abuseLimit) {
-		retryAfter = abuseLimit.GetRetryAfter()
-	} else {
+	if errIn == nil {
 		return false
 	}
 
-	githubNumRateLimitEncountered.WithLabelValues(s.name).Inc()
-	// Parse retry information from response headers, unless a Retry-After value was already provided.
-	// https://docs.github.com/en/rest/overview/resources-in-the-rest-api#exceeding-the-rate-limit
-	if retryAfter <= 0 && res != nil {
-		var err error
-		remaining, err = strconv.Atoi(res.Header.Get("x-ratelimit-remaining"))
-		if err != nil {
-			knownWait = false
+	rateLimitMu.RLock()
+	resumeTime := rateLimitResumeTime
+	rateLimitMu.RUnlock()
+
+	var retryAfter time.Duration
+	if resumeTime.IsZero() || time.Now().After(resumeTime) {
+		rateLimitMu.Lock()
+
+		var (
+			now = time.Now()
+
+			// GitHub has both primary (RateLimit) and secondary (AbuseRateLimit) errors.
+			limitType  string
+			rateLimit  *github.RateLimitError
+			abuseLimit *github.AbuseRateLimitError
+		)
+		if errors.As(errIn, &rateLimit) {
+			limitType = "primary"
+			rate := rateLimit.Rate
+			if rate.Remaining == 0 { // TODO: Will we ever receive a |RateLimitError| when remaining > 0?
+				retryAfter = rate.Reset.Sub(now)
+			}
+		} else if errors.As(errIn, &abuseLimit) {
+			limitType = "secondary"
+			retryAfter = abuseLimit.GetRetryAfter()
+		} else {
+			rateLimitMu.Unlock()
+			return false
 		}
 
-		resetTime, err := strconv.Atoi(res.Header.Get("x-ratelimit-reset"))
-		if err != nil || resetTime == 0 {
-			knownWait = false
-		} else if resetTime > 0 {
-			retryAfter = time.Duration(int64(resetTime)-time.Now().Unix()) * time.Second
+		jitter := time.Duration(rand.Intn(10)+1) * time.Second
```

Comment:
GitHub's API is buggy and will periodically reject requests that should have succeeded.
```
GET https://api.github.com/repos/org/repo: 403 API rate limit exceeded for user ID 32133502. If you reach out to GitHub Support for help, please include the request ID C8F0:12C9:34AADD1:6CC2F9A:653FC128. [rate limit was reset 0s ago]
GET https://api.github.com/repos/org/repo: 403 API rate limit exceeded for user ID 32133502. If you reach out to GitHub Support for help, please include the request ID C8F0:12C9:34AAE96:6CC3126:653FC128. [rate limit was reset 1s ago]
```

My solution to this is to add jitter, however...

1. I'm not sure if this is the best way to accomplish that (this specific implementation may be flawed)
2. It might be best to calculate jitter dynamically before calling `time.Sleep`

Reply:
The specific message `rate limit was reset 1s ago` [comes from the go-github package](https://github.com/google/go-github/blob/2cc47f9fa6b4d4997260c590388eecd19045c145/github/github.go#L1495-L1497) but I'm not sure if this is a bug with go-github. That error indicates that a 403 was received from GitHub with the header `x-remaining: 0`.

I suspect that it's a rare issue with GitHub's API that most people don't encounter because they don't frequently exhaust their ratelimit. In typical circumstances you'd have to wait an hour before resuming, however, scanning issues/prs triggers the secondary rate limit every minute.

In any case, adding a bit of jitter before retrying is probably something we should be doing regardless.